server/src/main/java/io/crate/execution/dml/RawIndexer.java

```diff
@@ -92,26 +100,54 @@ public ParsedDocument index(IndexItem item) throws IOException {
                 returnValues
             );
         });
-        Object[] insertValues = new Object[doc.size()];
+
+        currentRowValues = new Object[doc.size()];
         Iterator<Object> iterator = doc.values().iterator();
-        List<Reference> columns = indexer.columns();
-        for (int i = 0; i < insertValues.length; i++) {
+        List<Reference> columns = currentRowIndexer.columns();
+        for (int i = 0; i < currentRowValues.length; i++) {
             Reference reference = columns.get(i);
             Object value = iterator.next();
             DataType<?> type = reference.valueType();
             try {
-                insertValues[i] = type.implicitCast(value);
+                currentRowValues[i] = type.implicitCast(value);
             } catch (ClassCastException | IllegalArgumentException e) {
                 throw new ConversionException(value, type);
             }
         }
-        ParsedDocument parsedDoc = indexer.index(new IndexItem.StaticItem(
+
+        return currentRowIndexer.collectSchemaUpdates(new IndexItem.StaticItem(
             item.id(),
             item.pkValues(),
-            insertValues,
+            currentRowValues,
+            item.seqNo(),
+            item.primaryTerm()
```

Comment:
I BTW tried to not create a new Item but change signature to `..(ShardUpsertRequest.Item item)` and just mutate
```
item.insertValues(currentRowValues);
return currentRowIndexer.collectSchemaUpdates(item);
```
but then had multiple failures like 

```
Suppressed: org.opentest4j.AssertionFailedError: [All incoming requests on node [node_s0] should have finished] 
expected: 0L
 but was: 49320L
```

I think it's because original `ShardUpsertRequest` has only _raw and number of ref streamers must match with the number of value streamers.

Probably it's also a reason we create a new Item on mater branch for index(item).

Reply:
Yes if we would change the original item `insertValues` keyset, we'd have to change the requests `insertColumns` as well as they dictate how the values are streamed. I think this looks error-prone. So repeating the same operation of the replica as also done in master branch is the better approach afaik.